src/PIL/PpmImagePlugin.py

```diff
@@ -83,29 +97,21 @@ def _open(self):
             self.mode = rawmode = mode
 
         for ix in range(3):
-            while True:
-                while True:
-                    s = self.fp.read(1)
-                    if s not in b_whitespace:
-                        break
-                    if s == b"":
-                        raise ValueError("File does not extend beyond magic number")
-                if s != b"#":
-                    break
-                s = self.fp.readline()
-            s = int(self._token(s))
-            if ix == 0:
-                xsize = s
-            elif ix == 1:
-                ysize = s
+            try:  # check token sanity
+                token = int(self._read_token())
+            except ValueError:
+                raise SyntaxError("Non-decimal-ASCII found in header")
```

Comment:
(See https://github.com/python-pillow/Pillow/discussions/5104#discussioncomment-255699 for why this should remain a `ValueError`.)

And I don't think this exception message is necessarily better than what the original exception gives me. It doesn't tell me which bytes were the problem. I always prefer to see what value exactly it was it choked on. That makes it easier for me to see what I did wrong, or perhaps show enough info to file a PIL bug if it's something strange.

Or actually, it _does_ tell me, because this will show me 2 exceptions: first the original `ValueError` with its stack, then "During handling of the above exception, another exception occurred:" and the `SyntaxError` with its stack.

I don't think that amount of detail is useful for a PIL user. Re-raising might still be worth it to get a prettier message, but then I think it should do something like this to avoid the double stack:

```python
try:
    token = self._read_token()
    token = int(token)
except ValueError:
    raise ValueError(f"Better message with {token} in it") from None
```

Though this would need some additional work. If something is really wrong with the file or with PIL, this `token` could be any amount of data. The message for the built-in exception already truncates the value. This re-raised message does _not_ truncate the message, so a messed up token could give me a messed up exception message too...

So I'd say this isn't worth the trouble.

Reply:
But if you bring back the token size limit in some way then it's probably safe enough to do.

Reply:
I would use `_read_token` before the `try` because it can raise `ValueError` by itself if the file ends while reading the header.
It would be:
````python
token = self._read_token()  # if EOF is reached we get ValueError("Reached EOF while reading header")
try:
    token = int(token)
except ValueError:
    raise ValueError(f"Non-decimal-ASCII found in header: {token.decode('ascii')}") from None
````

Reply:
Right, yes.

But I don't think the `.decode('ascii')` should be there. If the token contains non-printable characters they'll disappear, or if it contains non-ASCII characters that'll raise a `UnicodeDecodeError`. I think it should just print the bytes value, unless `_read_token()` already raises its own exception if the token is not ASCII.

Reply:
https://github.com/python-pillow/Pillow/pull/5121/commits/5d0ad5e2e9a5065d69a00890342334f374277d1c addressed this by changing to `ValueError`, and then https://github.com/python-pillow/Pillow/pull/5121/commits/73fed77c0c58efd2ce11b8c0e96858bd03655163 added `from None`.

Reply:
Considering we'd had an issue in the past about multiple errors being raised (#4926), I think this is a step backwards. I've pushed a commit to remove this.